File: backend/app/schemas/workout.py
from __future__ import annotations

import uuid
from datetime import date, datetime

from pydantic import BaseModel, Field


# ── Create (nested input) ─────────────────────────────────────────────────────

class ExerciseSetCreate(BaseModel):
    set_number: int = Field(ge=1, le=100)
    reps: int = Field(ge=1, le=999)
    weight_kg: float = Field(ge=0.0, le=1000.0, default=0.0)


class WorkoutExerciseCreate(BaseModel):
    exercise_id: uuid.UUID
    sets: list[ExerciseSetCreate] = Field(min_length=1)


class WorkoutCreateRequest(BaseModel):
    session_date: date
    name: str | None = Field(default=None, max_length=100)
    notes: str | None = None
    is_shared: bool = False
    exercises: list[WorkoutExerciseCreate] = []


# ── Read (nested output) ──────────────────────────────────────────────────────

class ExerciseSetResponse(BaseModel):
    id: uuid.UUID
    set_number: int
    reps: int
    weight_kg: float


class WorkoutExerciseResponse(BaseModel):
    id: uuid.UUID
    exercise_id: uuid.UUID
    exercise_name: str
    muscle_group: str
    order_index: int
    sets: list[ExerciseSetResponse]


class WorkoutResponse(BaseModel):
    id: uuid.UUID
    session_date: date
    name: str | None
    notes: str | None
    is_shared: bool
    exercises: list[WorkoutExerciseResponse]
    created_at: datetime
    updated_at: datetime


# ── List (summary) ────────────────────────────────────────────────────────────

class WorkoutSummary(BaseModel):
    id: uuid.UUID
    session_date: date
    name: str | None
    is_shared: bool
    exercise_count: int
    total_sets: int
    created_at: datetime


class WorkoutListResponse(BaseModel):
    workouts: list[WorkoutSummary]
    total: int

File: frontend/types/workout.ts
export type MuscleGroup =
  | 'chest'
  | 'back'
  | 'shoulders'
  | 'biceps'
  | 'triceps'
  | 'legs'
  | 'core'
  | 'cardio'
  | 'full_body'
  | 'other'

export interface Exercise {
  id: string
  name: string
  muscle_group: MuscleGroup
  is_system: boolean
  created_by_user_id: string | null
  created_at: string
  updated_at: string
}

export interface ExerciseListResponse {
  exercises: Exercise[]
  total: number
}

export interface ExerciseCreateRequest {
  name: string
  muscle_group: MuscleGroup
}

export interface WorkoutSet {
  id: string
  set_number: number
  reps: number
  weight_kg: number
}

export interface WorkoutExercise {
  id: string
  exercise_id: string
  exercise_name: string
  muscle_group: MuscleGroup
  order_index: number
  sets: WorkoutSet[]
}

export interface WorkoutResponse {
  id: string
  session_date: string       // YYYY-MM-DD
  name: string | null
  notes: string | null
  is_shared: boolean
  exercises: WorkoutExercise[]
  created_at: string
  updated_at: string
}

export interface WorkoutSummary {
  id: string
  session_date: string
  name: string | null
  is_shared: boolean
  exercise_count: number
  total_sets: number
  created_at: string
}

export interface WorkoutListResponse {
  workouts: WorkoutSummary[]
  total: number
}

export interface WorkoutSetCreate {
  set_number: number
  reps: number
  weight_kg?: number
}

export interface WorkoutExerciseCreate {
  exercise_id: string
  sets: WorkoutSetCreate[]
}

export interface WorkoutCreateRequest {
  session_date: string
  name?: string | null
  notes?: string | null
  is_shared?: boolean
  exercises: WorkoutExerciseCreate[]
}

export interface LoggedExercise {
  exercise_id: string
  exercise_name: string
  muscle_group: string
  session_count: number
}

export interface LoggedExercisesResponse {
  exercises: LoggedExercise[]
  total: number
}

export interface ExerciseHistoryEntry {
  session_date: string  // YYYY-MM-DD
  max_weight_kg: number
  total_sets: number
  total_reps: number
  total_volume_kg: number
}

export interface ExerciseHistoryResponse {
  exercise_id: string
  exercise_name: string
  muscle_group: string
  entries: ExerciseHistoryEntry[]
  total_sessions: number
}

export interface PersonalRecord {
  exercise_id: string
  exercise_name: string
  muscle_group: string
  max_weight_kg: number
  achieved_on: string   // YYYY-MM-DD
  times_performed: number
}

export interface PersonalRecordsResponse {
  records: PersonalRecord[]
  total: number
}
